blockchain/blockchain.go

```diff
@@ -0,0 +1,198 @@
+	reason string
+}
+
+func (e ErrIncompatibleBlockAndStateUpdate) Error() string {
+	return fmt.Sprintf("incompatible block and state update: %v", e.reason)
+}
+
+type ErrIncompatibleBlock struct {
+	reason string
+}
+
+func (e ErrIncompatibleBlock) Error() string {
+	return fmt.Sprintf("incompatible block: %v", e.reason)
+}
+
+// BlockDbKey creates appends block number and hash to create a db key.
+// Todo: make private after Blockchain.BlockByNumber & BlockchainBlockByHash are implemented.
+type BlockDbKey struct {
+	Number uint64
+	Hash   *felt.Felt
+}
+
+func (k *BlockDbKey) MarshalBinary() ([]byte, error) {
+	var numB [8]byte
+	binary.BigEndian.PutUint64(numB[:], k.Number)
+	return db.Blocks.Key(numB[:], k.Hash.Marshal()), nil
+}
+
+func (k *BlockDbKey) UnmarshalBinary(data []byte) error {
+	if len(data) != 41 {
+		return errors.New("key should be 41 bytes long")
+	}
+
+	if data[0] != byte(db.Blocks) {
+		return errors.New("wrong prefix")
+	}
+
+	k.Number = binary.BigEndian.Uint64(data[1:9])
+	k.Hash = new(felt.Felt).SetBytes(data[9:41])
+	return nil
+}
+
+// Blockchain is responsible for keeping track of all things related to the StarkNet blockchain
+type Blockchain struct {
+	l *sync.RWMutex
+
+	head     *core.Block
+	network  utils.Network
+	database db.DB
+}
+
+func NewBlockchain(database db.DB, network utils.Network) (*Blockchain, error) {
+	chain := &Blockchain{
+		l:        &sync.RWMutex{},
+		database: database,
+		network:  network,
+	}
+	txn := chain.database.NewTransaction(false)
+	headBlockBin, err := txn.Get(db.HeadBlock.Key())
+	if errors.Is(err, db.ErrKeyNotFound) {
+		return chain, nil
+	}
+
+	headBlock := new(core.Block)
+	if err = gob.NewDecoder(bytes.NewReader(headBlockBin)).Decode(headBlock); err != nil {
+		return nil, err
+	}
+	chain.head = headBlock
+	return chain, nil
+}
+
+// Height returns the latest block height. If blockchain is empty nil is returned.
+func (b *Blockchain) Height() *uint64 {
+	b.l.RLock()
+	defer b.l.RUnlock()
+	if b.head != nil {
+		h := b.head.Number
+		return &h
+	}
+	return nil
+}
+
+// Store takes a block and state update and performs sanity checks before putting in the database.
+func (b *Blockchain) Store(block *core.Block, stateUpdate *core.StateUpdate) error {
+	if err := b.VerifyBlock(block, stateUpdate); err != nil {
+		return err
+	}
+	if err := state.NewState(b.database).Update(stateUpdate); err != nil {
+		return err
+	}
+
+	key := &BlockDbKey{block.Number, block.Hash}
+	bKey, err := key.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	buffer := new(bytes.Buffer)
+	if err = gob.NewEncoder(buffer).Encode(block); err != nil {
+		return err
+	}
+	blockBinary := append([]byte{}, buffer.Bytes()...)
+
+	if err = b.database.Update(func(txn db.Transaction) error {
+		if err = txn.Set(db.HeadBlock.Key(), blockBinary); err != nil {
+			return err
+		}
+		return txn.Set(bKey, blockBinary)
+	}); err != nil {
+		return err
+	}
+
+	b.l.Lock()
+	b.head = block
+	b.l.Unlock()
+	return nil
+}
```

Comment:
`Store()` as a whole should be atomic and reentrant. 

It is not atomic because, there are multiple points that we could error out of this function when we have already altered some state and we don't rollback any of those changes. For example, if gob.Encode() fails, we have already called the state.Update() and modified the DB. We should be creating and using only a single DB txn thru out the entire `Store()` call.

It is not reentrant because, for example, we could verify a block but before we can apply the state or store the block, another call to `Store()` could change the `HEAD`.

Reply:
Done